File: Python/第二章/11-rag-evaluation/run_evaluation.py
"""运行固定 RAG 评估集并打印 Recall@K、MRR 与 Faithfulness。"""

import sys
from typing import Any

from evaluation_cases import evaluation_cases
from rag_evaluator import (
    diagnose,
    evaluate_faithfulness,
    evaluate_retrieval,
    mean,
)


TOP_K = 3


def run_evaluation(mode: str) -> list[dict[str, Any]]:
    """按照指定模式运行全部评估案例。"""
    if mode not in {"retrieval", "full"}:
        raise ValueError("运行模式只能是 retrieval 或 full。")

    results: list[dict[str, Any]] = []

    for evaluation_case in evaluation_cases:
        # Recall@K 和 RR 完全由固定标注与检索结果计算，不需要调用模型。
        retrieval = evaluate_retrieval(evaluation_case, TOP_K)

        # full 模式才调用评估模型判断 Faithfulness。
        faithfulness_result = (
            evaluate_faithfulness(evaluation_case)
            if mode == "full"
            else None
        )
        faithfulness = (
            faithfulness_result["score"]
            if faithfulness_result is not None
            else None
        )

        result = {
            "id": evaluation_case["id"],
            "name": evaluation_case["name"],
            "question": evaluation_case["question"],
            **retrieval,
            "faithfulness": faithfulness,
            "claims": (
                faithfulness_result["claims"]
                if faithfulness_result is not None
                else []
            ),
            "diagnosis": diagnose(
                {
                    **retrieval,
                    "faithfulness": faithfulness,
                }
            ),
        }
        results.append(result)

    return results


def print_results(
    results: list[dict[str, Any]], mode: str
) -> None:
    """打印每个案例和整个评估集的指标。"""
    print(f"评估模式：{'完整评估' if mode == 'full' else '仅检索指标'}")
    print(f"TopK：{TOP_K}")

    for result in results:
        print(f"\n================ {result['name']} ================")
        print(f"问题：{result['question']}")

        recall = result["recall"]
        recall_text = "N/A" if recall is None else f"{recall:.6f}"
        print(f"Recall@{TOP_K}：{recall_text}")
        print(f"RR@{TOP_K}：{result['reciprocalRank']:.6f}")

        if result["faithfulness"] is not None:
            print(f"Faithfulness：{result['faithfulness']:.6f}")
            print("Claim 判断：")

            for claim in result["claims"]:
                status = "支持" if claim["supported"] else "不支持"
                print(f"- [{status}] {claim['claim']}")

        print(f"定位建议：{result['diagnosis']}")

    # 没有人工相关性标注的案例不参与 Mean Recall@K 统计。
    valid_recall_values = [
        result["recall"]
        for result in results
        if result["recall"] is not None
    ]
    mean_recall = mean(valid_recall_values)
    mrr = mean([result["reciprocalRank"] for result in results])

    print("\n================ 整体指标 ================")
    print(f"Mean Recall@{TOP_K}：{mean_recall:.6f}")
    print(f"MRR@{TOP_K}：{mrr:.6f}")

    if mode == "full":
        mean_faithfulness = mean(
            [
                result["faithfulness"]
                for result in results
                if result["faithfulness"] is not None
            ]
        )
        print(f"Mean Faithfulness：{mean_faithfulness:.6f}")


def main() -> None:
    """读取命令行模式，运行评估并输出结果。"""
    mode = sys.argv[1] if len(sys.argv) > 1 else "retrieval"
    results = run_evaluation(mode)
    print_results(results, mode)


if __name__ == "__main__":
    main()

File: Node/第二章/11-rag-evaluation/run-evaluation.js
import { evaluationCases } from './evaluation-cases.js'
import {
	diagnose,
	evaluateFaithfulness,
	evaluateRetrieval,
	mean
} from './rag-evaluator.js'

const TOP_K = 3
const mode = process.argv[2] ?? 'retrieval'

if (!['retrieval', 'full'].includes(mode)) {
	throw new Error('运行模式只能是 retrieval 或 full。')
}

const results = []

for (const evaluationCase of evaluationCases) {
	const retrieval = evaluateRetrieval(evaluationCase, TOP_K)
	const faithfulnessResult =
		mode === 'full'
			? await evaluateFaithfulness(evaluationCase)
			: null

	const result = {
		id: evaluationCase.id,
		name: evaluationCase.name,
		question: evaluationCase.question,
		...retrieval,
		faithfulness: faithfulnessResult?.score ?? null,
		claims: faithfulnessResult?.claims ?? [],
		diagnosis: diagnose({
			...retrieval,
			faithfulness: faithfulnessResult?.score ?? null
		})
	}

	results.push(result)
}

console.log(`评估模式：${mode === 'full' ? '完整评估' : '仅检索指标'}`)
console.log(`TopK：${TOP_K}`)

for (const result of results) {
	console.log(`\n================ ${result.name} ================`)
	console.log(`问题：${result.question}`)
	console.log(
		`Recall@${TOP_K}：${
			result.recall === null ? 'N/A' : result.recall.toFixed(6)
		}`
	)
	console.log(`RR@${TOP_K}：${result.reciprocalRank.toFixed(6)}`)

	if (result.faithfulness !== null) {
		console.log(`Faithfulness：${result.faithfulness.toFixed(6)}`)
		console.log('Claim 判断：')

		for (const claim of result.claims) {
			console.log(
				`- [${claim.supported ? '支持' : '不支持'}] ${claim.claim}`
			)
		}
	}

	console.log(`定位建议：${result.diagnosis}`)
}

const validRecallValues = results
	.map((result) => result.recall)
	.filter((value) => value !== null)
const meanRecall = mean(validRecallValues)
const mrr = mean(results.map((result) => result.reciprocalRank))

console.log('\n================ 整体指标 ================')
console.log(`Mean Recall@${TOP_K}：${meanRecall.toFixed(6)}`)
console.log(`MRR@${TOP_K}：${mrr.toFixed(6)}`)

if (mode === 'full') {
	const meanFaithfulness = mean(
		results
			.map((result) => result.faithfulness)
			.filter((value) => value !== null)
	)
	console.log(`Mean Faithfulness：${meanFaithfulness.toFixed(6)}`)
}
